Document `CHAINERX_CUDNN_USE_CUPY` (#7574)
chainer/chainer: docs/source/chainerx/install/index.rst

----------------------

The following environment variables are available for building ChainerX from source.

=========================== ========================================================================================================
Environment variable        Description
=========================== ========================================================================================================
``CHAINER_BUILD_CHAINERX``  ``1`` to build the ``chainerx`` package along with ``chainer``. ``0`` to skip. Default is ``0``.
``CHAINERX_BUILD_CUDA``     ``1`` to build ``chainerx`` with CUDA support. ``0`` to skip. Default is ``0``.
                            See also :ref:`CUDA support <chainerx-install-cuda-support>` section below.
``CHAINERX_ENABLE_BLAS``    ``1`` to make BLAS enabled. ``0`` to disabled. Default is ``1``.
=========================== ========================================================================================================

Simply run ``pip install --pre chainer`` after configuring the above environment variables.
See :ref:`Examples <chainerx-install-cuda-support>` below.

.. _chainerx-install-cuda-support:

CUDA support
------------

> When installing with the CUDA support, you also need to specify the cuDNN installation path since CUDA without cuDNN is currently not supported.

Reviewer: ```suggestion
When installing with the CUDA support, you also need to specify the cuDNN installation path since current implementation of ChainerX requires cuDNN.
```

Just that I don't like `not not` expression.

Author: There's virtually no possibility to support CUDA without cuDNN. I just removed the clause as I feel it's redundant. How do you think?

Reviewer: I like short explanation.

                            replaced.
=========================== ========================================================================================================

To support the :ref:`NumPy/CuPy fallback <chainerx-tutorial-numpy-cupy-fallback>` mechanism, currently ChainerX with the CUDA support requires CuPy to be installed together.

.. seealso::

    `CuPy installation guide <https://docs-cupy.chainer.org/en/stable/install.html>`_

.. _chainerx-install-examples:

Examples
--------

Install ChainerX without CUDA support:

.. code-block:: console

    $ export CHAINER_BUILD_CHAINERX=1
    $ export MAKEFLAGS=-j8  # Using 8 parallel jobs.
    $ pip install --pre chainer

Install ChainerX depending on CuPy wheel distribution:

.. code-block:: console

    $ pip install --pre cupy_cuda101  # Note: Choose the proper CUDA SDK version number.
    $ export CHAINER_BUILD_CHAINERX=1
    $ export CHAINERX_BUILD_CUDA=1
    $ export CHAINERX_CUDNN_USE_CUPY=1
    $ export MAKEFLAGS=-j8  # Using 8 parallel jobs.
    $ pip install --pre chainer

Install ChainerX with CuPy built from source:

> .. code-block:: console

Reviewer: Do we still need this install instruction below?
I think `CHAINERX_CUDNN_USE_CUPY` should be the next standard way to build with ChainerX.

Author: I assume non-negligible proportion of developers build CuPy from source.
In fact, it's adviced to align CuPy version to that of Chainer: those who build ChainerX from master branch should also use CuPy master branch. I think this example is informative for those developers.

Reviewer: Seems reasonable

Configure the following environment variables before installing Chainer.

> 

Reviewer: Shouldn't this explanation updated too since `CHAINERX_CUDNN_USE_CUPY` has higher priority?

Author: How do you think should it be updated? Just the order?

I don't think `CHAINERX_CUDNN_USE_CUPY` has a higer priority. It's just a "convenience" function for a limited number of audience.

Reviewer: When `CHAINERX_CUDNN_USE_CUPY` is set it's not required any more so it should be noted.
I think currently most users would install cupy with chainerx so I feel it's a not limited thing.

Author: Currently ChainerX depends on CuPy only temporarily. I don't want to document as if it were a norm.
Also this option has unobvious pitfalls as written in the docs.

Reviewer: As long as ChainerX is depending on CuPy I like the convenience function more since installing cuDNN is kind a difficult step in the setup.
And I think the checks in the cmake script is enough to know the reason:
https://github.com/chainer/chainer/blob/f8db3447e84634f29fbce37d5c253eed4cb1910f/chainerx_cc/CMakeLists.txt#L84-L105

Author: > And I think the checks in the cmake script is enough to know the reason:

Could you elabolate what this means?

I do think `CHAINERX_CUDNN_USE_CUPY` is not the "authentic" method that anyone can use. Do you agree with that?
The "authentic" one (`CUDNN_ROOT_DIR`) should come first IMO.
